x-pack/plugins/security/server/plugin.ts

```diff
@@ -12,36 +13,77 @@ import {
   Logger,
   PluginInitializerContext,
   RecursiveReadonly,
+  SavedObjectsLegacyService,
+  LegacyRequest,
 } from '../../../../src/core/server';
 import { deepFreeze } from '../../../../src/core/utils';
-import { XPackInfo } from '../../../legacy/plugins/xpack_main/server/lib/xpack_info';
-import { setupAuthentication, Authentication } from './authentication';
+import { SpacesPluginSetup } from '../../spaces/server';
+import { PluginSetupContract as FeaturesSetupContract } from '../../features/server';
+import { LicensingPluginSetup } from '../../licensing/server';
+import { CapabilitiesModifier } from '../../../../src/legacy/server/capabilities';
+
+import { Authentication, setupAuthentication } from './authentication';
+import { Authorization, setupAuthorization } from './authorization';
 import { createConfig$ } from './config';
 import { defineRoutes } from './routes';
+import { SecurityLicenseService, SecurityLicense } from './licensing';
+import { setupSavedObjects } from './saved_objects';
+import { SecurityAuditLogger } from './audit';
+
+export type SpacesService = Pick<
+  SpacesPluginSetup['spacesService'],
+  'getSpaceId' | 'namespaceToSpaceId'
+>;
+
+export type FeaturesService = Pick<FeaturesSetupContract, 'getFeatures'>;
 
 /**
  * Describes a set of APIs that is available in the legacy platform only and required by this plugin
  * to function properly.
  */
 export interface LegacyAPI {
-  xpackInfo: Pick<XPackInfo, 'isAvailable' | 'feature'>;
   isSystemAPIRequest: (request: KibanaRequest) => boolean;
+  capabilities: { registerCapabilitiesModifier: (provider: CapabilitiesModifier) => void };
+  kibanaIndexName: string;
   cspRules: string;
+  savedObjects: SavedObjectsLegacyService<KibanaRequest | LegacyRequest>;
+  auditLogger: {
+    log: (eventType: string, message: string, data?: Record<string, unknown>) => void;
+  };
 }
 
 /**
  * Describes public Security plugin contract returned at the `setup` stage.
  */
 export interface PluginSetupContract {
   authc: Authentication;
+  authz: Pick<Authorization, 'actions' | 'checkPrivilegesWithRequest' | 'mode'>;
+
+  /**
+   * If Spaces plugin is available it's supposed to register its SpacesService with Security plugin
+   * so that Security can get space ID from the URL or namespace. We can't declare optional dependency
+   * to Spaces since it'd result into circular dependency between these two plugins and circular
+   * dependencies aren't supported by the Core. In the future we have to get rid of this implicit
+   * dependency.
+   * @param service Spaces service exposed by the Spaces plugin.
+   */
+  registerSpacesService: (service: SpacesService) => void;
```

Comment:
__note:__ that is the workaround for the circular dependency I mentioned previously. We could potentially make it more generic, but I'd like to discuss/see-how-it-goes first.

Reply:
> For a short-term solution, I think this is reasonable. I don't think it should become the long-term solution though. During the security plugin's setup lifecycle event, we're exposing authz and if some other plugin happens to use anything that depends on getSpacesService before the spaces plugin has a chance to call registerSpacesService, we'll be executing "incorrect" logic.

Yeah, even if we can "solve" this via moving methods that depend on space service to `start` contract I don't see it as a good enough long-term solution.

> For the short-term, can we add logic to throw an error if registerSpacesService is called after there are any calls to getSpacesService? This way, it'll at least be obvious if this happens to occur.

That's a good idea, I'm not sure why I didn't do it in the same way as `getLegacyAPI` that throws if it's called before legacy API is registered :man_shrugging: